backend/backend.c

```diff
@@ -91,37 +91,46 @@ struct wlr_backend *wlr_backend_autocreate(struct wl_display *display) {
 		return NULL;
 	}
 
-	int gpu = wlr_session_find_gpu(session);
-	if (gpu == -1) {
-		wlr_log(L_ERROR, "Failed to open DRM device");
-		goto error_session;
-	}
-
 	backend = wlr_multi_backend_create(session);
 	if (!backend) {
-		goto error_gpu;
+		goto error_session;
 	}
 
 	struct wlr_backend *libinput = wlr_libinput_backend_create(display, session);
 	if (!libinput) {
 		goto error_multi;
 	}
 
-	struct wlr_backend *drm = wlr_drm_backend_create(display, session, gpu);
-	if (!drm) {
-		goto error_libinput;
+	wlr_multi_backend_add(backend, libinput);
+
+	int gpus[8];
+	size_t num_gpus = wlr_session_find_gpus(session, 8, gpus);
+	struct wlr_backend *primary_drm = NULL;
+	wlr_log(L_INFO, "Found %zu GPUs", num_gpus);
+
+	for (size_t i = 0; i < num_gpus; ++i) {
+		struct wlr_backend *drm = wlr_drm_backend_create(display, session, gpus[i]);
+		if (!drm) {
+			wlr_log(L_ERROR, "Failed to open DRM device");
+			continue;
+		}
+
+		if (!primary_drm) {
+			primary_drm = drm;
+		}
+
+		wlr_multi_backend_add(backend, drm);
+	}
+
+	if (!primary_drm) {
+		wlr_log(L_ERROR, "Failed to open any DRM device");
+		goto error_multi;
```

Comment:
Something that might be useful in the future is to specify a list of GPUs to use - then maybe you can run a different compositor on each GPU, for example.

Reply:
Yes, being able to explicitly pick the GPUs would be helpful.
Maybe a colon-separated environment variable would be helpful, where the first one specified is the primary GPU:
WLR_DRM_DEVICES=/dev/dri/card0:/dev/dri/card3